fedimintd/src/lib.rs

```diff
@@ -0,0 +1,61 @@
+use ring::aead::{Aad, LessSafeKey, Nonce, UnboundKey, NONCE_LEN};
+
+use ring::{aead, digest, pbkdf2};
+use std::fs;
+use std::num::NonZeroU32;
+use std::path::PathBuf;
+const ITERATIONS: Option<NonZeroU32> = NonZeroU32::new(100_000);
+// This value was generated from a secure PRNG.
+const SALT: [u8; 16] = [
+    95, 90, 223, 174, 131, 17, 64, 89, 14, 42, 26, 247, 94, 71, 125, 13,
+];
+
+pub fn encrypted_write(mut data: Vec<u8>, key: &LessSafeKey, nonce: Nonce, file: PathBuf) {
+    key.seal_in_place_append_tag(nonce, Aad::empty(), &mut data)
+        .expect("encrypted");
+    let encrypted_hex = hex::encode(data);
+    fs::write(file, &encrypted_hex).expect("Can't write file.");
+}
+
+pub fn encrypted_read(key: &LessSafeKey, nonce: Nonce, file: PathBuf) -> Vec<u8> {
+    let encrypted_hex = fs::read_to_string(file).expect("Can't read file.");
+    let mut encrypted_bytes = hex::decode(encrypted_hex).expect("not hex encoded");
+    key.open_in_place(nonce, Aad::empty(), &mut encrypted_bytes)
+        .expect("decrypts");
+    encrypted_bytes.truncate(encrypted_bytes.len() - key.algorithm().tag_len());
+    encrypted_bytes
+}
+
+/// Generate an key from the user password using PBKDF2
+///
+/// ChaCha20-Poly1305 is used to encrypt and authenticate the configs
+pub fn get_key(password: Option<String>) -> LessSafeKey {
+    let password = match password {
+        None => rpassword::prompt_password("Enter a password to encrypt configs: ").unwrap(),
+        Some(password) => {
+            println!("WARNING: Passing in a password from the command line may be less secure!");
+            password
+        }
+    };
+
+    let mut key = [0u8; digest::SHA256_OUTPUT_LEN];
+    let algo = pbkdf2::PBKDF2_HMAC_SHA256;
+    pbkdf2::derive(
+        algo,
+        ITERATIONS.unwrap(),
+        &SALT,
+        password.as_bytes(),
+        &mut key,
+    );
+    let key = UnboundKey::new(&aead::CHACHA20_POLY1305, &key).expect("created key");
+    LessSafeKey::new(key)
+}
+
+pub fn tls_nonce() -> Nonce {
+    Nonce::assume_unique_for_key([0; NONCE_LEN])
+}
+
+// FIXME - encrypting different versions of the configs will require different nonces to be secure
+pub fn config_nonce() -> Nonce {
+    Nonce::assume_unique_for_key([1; NONCE_LEN])
+}
```

Comment:
Is "nonce" here a different term for initialization vector? Hard coded nonces are a big red flag imo. Can we deterministically derive it from the key/content somehow? Is it used for en/decryption or just authentication?

Reply:
Ok, do you have a plan for fixing it? Password re-use would trigger the same, right? So I'd like to fix this ASAP if we land it like that.

Reply:
> In the case we need to generate new versions of the configs we could have an auto-incrementing counter stored in the config file.

But if the config is encrypted we wouldn't know the counter at decryption time, right? If we don't need to know it then we should just hash the config and use that as the nonce.

Reply:
We could store the incrementing nonce in the config file itself as plaintext.  I could make that change now I suppose.